### DISCUSSION TITLE
Testing with the `PersistentContainer` Dependencies Addition

### DISCUSSION THREAD

**Discussion by Nomad00** [tgrapperon/swift-dependencies-additions#49] [Q&A] [answered]
Greetings all 👋 

I've run into an issue with an unimplemented dependancy and I am unsure if I'm using it wrong or if something needs to be tweaked with `PersistentContainer`'s `testValue`.

Given this Reducer:
```swift
struct ItemListFeature: Reducer {
    @Dependency(\.persistentContainer)
    var persistentContainer
    
    struct State: Equatable {
        var items: Item.FetchedResults = .empty
    }
    
    enum Action: Equatable { }
    
    func reduce(into state: inout State, action: Action) -> Effect<Action> {
        .none
    }
}
```

I created the following test:
```swift
@MainActor
class ItemListFeatureTests: XCTestCase {
   func testBasics() async {
        let store = TestStore(
            initialState: ItemListFeature.State(),
            reducer: ItemListFeature())
        
        store.dependencies.persistentContainer = .default(
            inMemory: true
        )
    }
}
```
I get the following error:

> testBasics(): Unimplemented: @Dependency(\.persistentContainer.newBackgroundContext) …
>  Defined at:
>    PersistentContainerDependency/PersistentContainer.swift:28

I don't expect to have to override `newBackgroundContext`. If I am, it is unclear to me how to properly override it myself. Is anyone able to set me straight here? Thanks!

**Comment by tgrapperon** [comment, marked as answer]
Hey @Nomad00!
I just tried your code in a standalone project, and I'm not able to reproduce the issue.
It is maybe possible that your host application is hitting this dependency while testing. This kind of issue can also happen if you're statically linking `Dependencies` in your test target.
Have you checked this [section of the documentation](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/testing/#Testing-gotchas)?
In any case, you shouldn't have to implement individual endpoints, `.default(inMemory:)` is a fully fledged persistent store.

**Reply by Nomad00** [reply]
That you so much for the response @tgrapperon ! I'm pretty sure I have the linking right, but that is exactly the type of problem I was expecting. I'm going to make a fresh project and see what happens.

**Reply by Nomad00** [reply]
And did not happen in a fresh project. Thanks again @tgrapperon , I must have something funky with my host application. Thanks for listening / reading!

**Reply by tgrapperon** [reply]
One thing you can try is using the trailing closure to set your dependencies up in `TestStore`:
```swift
let store = TestStore(
  initialState: ItemListFeature.State(),
  reducer: ItemListFeature()) {
  $0.persistentContainer = .default(inMemory: true)
}
```
This is slightly different from setting it later because the dependency will be set when you create your state (but it would be rather odd to touch a `newBackgroundContext` there).
Please let us know your findings, it is also possible that something can be improved in this area.

**Comment by Nomad00** [comment]
Hey @tgrapperon ! Apologies that I'm only getting to this sporadically, but I started a repo where I'm hoping to use `PersistentContainer` with TCA and have test coverage of basic operations. If you'd like to take a look, feel free https://github.com/Nomad00/TestingPersistentContainer.
